File: javascript/src/memento/memento.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { Caretaker } from "./caretaker.js";
import { Originator } from "./originator.js";

describe("Memento", () => {
  it("saved states are restored most recent first", () => {
    const originator = new Originator();
    const caretaker = new Caretaker();
    originator.setState("one");
    caretaker.save(originator);
    originator.setState("two");
    caretaker.save(originator);
    originator.setState("three");
    assert.equal(caretaker.undo(originator), true);
    assert.equal(originator.getState(), "two");
    assert.equal(caretaker.undo(originator), true);
    assert.equal(originator.getState(), "one");
  });

  it("undo with nothing saved leaves the originator alone", () => {
    const originator = new Originator();
    originator.setState("x");
    assert.equal(new Caretaker().undo(originator), false);
    assert.equal(originator.getState(), "x");
  });

  it("a memento captures the value at save time", () => {
    const originator = new Originator();
    originator.setState("before");
    const memento = originator.createMemento();
    originator.setState("after");
    originator.restore(memento);
    assert.equal(originator.getState(), "before");
  });
});

File: java/src/test/java/com/penapereira/patterns/memento/MementoTest.java
package com.penapereira.patterns.memento;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class MementoTest {

    @Test
    void savedStatesAreRestoredMostRecentFirst() {
        Originator originator = new Originator();
        Caretaker caretaker = new Caretaker();
        originator.setState("one");
        caretaker.save(originator);
        originator.setState("two");
        caretaker.save(originator);
        originator.setState("three");
        assertTrue(caretaker.undo(originator));
        assertEquals("two", originator.getState());
        assertTrue(caretaker.undo(originator));
        assertEquals("one", originator.getState());
    }

    @Test
    void undoWithNothingSavedLeavesTheOriginatorAlone() {
        Originator originator = new Originator();
        originator.setState("x");
        assertFalse(new Caretaker().undo(originator));
        assertEquals("x", originator.getState());
    }

    @Test
    void aMementoCapturesTheValueAtSaveTime() {
        Originator originator = new Originator();
        originator.setState("before");
        Originator.Memento memento = originator.createMemento();
        originator.setState("after");
        originator.restore(memento);
        assertEquals("before", originator.getState());
    }
}
